src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_TupleBinaryOperator.cs

```diff
@@ -0,0 +1,355 @@
+            return EvaluateSideEffectingArgumentToTemp(VisitExpression(expr), initEffects, ref temps);
+        }
+
+        private BoundExpression RewriteTupleOperator(TupleBinaryOperatorInfo @operator,
+            BoundExpression left, BoundExpression right, TypeSymbol boolType,
+            TupleOperatorSideEffectsAndTemps initialEffectsAndTemps, BinaryOperatorKind operatorKind)
+        {
+            if (@operator.IsSingle())
+            {
+                return RewriteTupleSingleOperator((TupleBinaryOperatorInfo.Single)@operator, left, right, boolType, operatorKind);
+            }
+            else
+            {
+                return RewriteTupleNestedOperators((TupleBinaryOperatorInfo.Multiple)@operator, left, right, boolType, initialEffectsAndTemps, operatorKind);
+            }
+        }
+
+        private BoundExpression RewriteTupleNestedOperators(TupleBinaryOperatorInfo.Multiple operators, BoundExpression left, BoundExpression right,
+            TypeSymbol boolType, TupleOperatorSideEffectsAndTemps initialEffectsAndTemps, BinaryOperatorKind operatorKind)
+        {
+            // If either left or right is nullable, produce:
+            //
+            //      // outer sequence
+            //      leftHasValue = left.HasValue; (or true if !leftNullable)
+            //      leftHasValue = right.HasValue (or true if !rightNullable)
+            //          ? leftHasValue ? ... inner sequence ... : true/false
+            //          : false/true
+            //
+            // where inner sequence is:
+            //      leftValue = left.GetValueOrDefault(); (or left if !leftNullable)
+            //      rightValue = right.GetValueOrDefault(); (or right if !rightNullable)
+            //      ... logical expression using leftValue and rightValue ...
+            //
+            // and true/false and false/true depend on operatorKind (== vs. !=)
+            //
+            // and temps are created an initialized with early side-effects for:
+            //  - elements from left and right, in case of a tuple literal
+            //  - left and right, in case of a tuple or nullable tuple type
+            //
+            // But if neither is nullable, then just produce the inner sequence.
+            //
+            // Note: all the temps are created in a single bucket (rather than different scopes of applicability) for simplicity
+
+            // PROTOTYPE(tuple-equality) Consider if optimizations from TrivialLiftedComparisonOperatorOptimizations can be applied
+
+            var outerEffects = ArrayBuilder<BoundExpression>.GetInstance();
+            var innerEffects = ArrayBuilder<BoundExpression>.GetInstance();
+
+            BoundExpression leftHasValue;
+            BoundExpression leftValue;
+
+            // Note: left and right are either temps or `null`, so we don't have detailed information to tell us a nullable always has a value
```

Comment:
>// Note: left and right are either temps or `null`, so we don't have detailed information to tell us a nullable always has a value [](start = 12, length = 130)

Before after we visit the argument, but before we store we store it into a temp we can get this information, we even can unwrap it if it is never null. Let's try to follow-up in a separate PR